File: WolfDen/Game/Behaviours/CaptureBehaviour/NormalCaptureBehaviourTemplate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using WolfDen.Game;
using WolfDen.Game.Interfaces;

namespace WolfDen.Game.Templates
{
    public class NormalCaptureBehaviourTemplate : IBehaviourTemplate
    {
        public string Class { get { return GetType().Name; } }
        public int BaseCaptureStrength { get; set; }
        public int CaptureCompletion { get; set; }



        public void AddBehaviour(Unit unit)
        {
            var moveBehaviour = unit.Behaviours.Values.OfType<IMoveBehaviour>().FirstOrDefault();
            if (moveBehaviour == null)
                throw new Exception("No move behaviour found while generating capture behaviour. Make sure it comes earlier in the list.");
            var iHealth = unit.Behaviours.Values.OfType<IHealth>().FirstOrDefault();
            if (iHealth == null)
                throw new Exception("No IHealth (or combatbehaviour) found while generating capture behaviour. Make sure it comes earlier in the list.");
            var captureBehaviour = new NormalCaptureBehaviour(unit.Game, unit, BaseCaptureStrength, CaptureCompletion, moveBehaviour, iHealth);
            captureBehaviour.Id = unit.Id + Class;
            unit.Behaviours.Add(captureBehaviour.Id, captureBehaviour);
        }
    }
}

File: WolfDen/wwwroot/ts/Behaviours/CaptureBehaviour/NormalCaptureBehaviourTemplate.ts
/// <reference path="../../Unit.ts" />

namespace TurnBasedBoardGame {
    export class NormalCaptureBehaviourTemplate implements IBehaviourTemplate {
        BaseCaptureStrength: number = 0;
        CaptureCompletion: number = 0;
        Class: string;

        AddBehaviour(unit: Unit): void {
            let self = this;
            let captureBehaviour = new NormalCaptureBehaviour();
            Object.assign(captureBehaviour, self);
            let moveBehaviour = null;
            for (let key in unit.Behaviours)
                if (IsIMoveBehaviour(unit.Behaviours[key]))
                    moveBehaviour = unit.Behaviours[key];
            let combatBehaviour = null;
            for (let key in unit.Behaviours)
                if (IsIHealth(unit.Behaviours[key]))
                    combatBehaviour = unit.Behaviours[key];
            captureBehaviour.MoveBehaviour = moveBehaviour;
            captureBehaviour.HealthStats = combatBehaviour;
            captureBehaviour.Unit = unit;
            captureBehaviour.Game = unit.Game;  
            captureBehaviour.Id = unit.Id + self.Class;
            captureBehaviour.Initialize();
            
            unit.Behaviours[captureBehaviour.Id] = captureBehaviour;
        }
    }
}